org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/SynchronizeGradleBuildsOperation.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2016 the original author or authors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.eclipse.buildship.core.workspace.internal;
+
+import java.util.Set;
+
+import org.gradle.tooling.CancellationTokenSource;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+
+import com.gradleware.tooling.toolingmodel.OmniEclipseProject;
+import com.gradleware.tooling.toolingmodel.repository.FetchStrategy;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.OperationCanceledException;
+import org.eclipse.core.runtime.SubMonitor;
+
+import org.eclipse.buildship.core.configuration.BuildConfiguration;
+import org.eclipse.buildship.core.workspace.GradleBuild;
+import org.eclipse.buildship.core.workspace.GradleBuilds;
+import org.eclipse.buildship.core.workspace.ModelProvider;
+import org.eclipse.buildship.core.workspace.NewProjectHandler;
+
+/**
+ * Synchronizes each of the given Gradle builds with the workspace.
+ */
+public final class SynchronizeGradleBuildsOperation {
+
+    private final ImmutableSet<GradleBuild> builds;
+    private final NewProjectHandler newProjectHandler;
+
+    private SynchronizeGradleBuildsOperation(Set<GradleBuild> builds, NewProjectHandler newProjectHandler) {
+        this.builds = ImmutableSet.copyOf(builds);
+        this.newProjectHandler = Preconditions.checkNotNull(newProjectHandler);
+    }
+
+    Set<GradleBuild> getBuilds() {
+        return this.builds;
+    }
+
+    protected void run(CancellationTokenSource tokenSource, IProgressMonitor monitor) throws Exception {
```

Comment:
⭕️ Would it make sense to have some common interface like `ToolingApiRunnable` for this? Then we could have `CorePlugin.run(tapiRunnable)`, which would take care of getting a token source for you.

Reply:
I've introduced `ToolingApiOperation` both in the synchronous invocations and in the jobs.